Disable WooPay for suspended and rejected accounts

asumaran opened the pull request:
Closes https://github.com/Automattic/woopay/issues/2641

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

Add checks to `WC_Payments_Features::is_woopay_eligible()` to return `false` in case the merchant account is suspended or rejected.

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

### Testing instructions
---
> [!Warning]  
> Rejecting a merchant account is irreversible. It will inform Stripe about the rejection and it can not be undone.

Since I didn't find a way to reverse a rejection I'm sharing another way to simulate a rejected account.

**a) Reject the account using the MC tool.**
  - In the WCPay Server’ admin area:
  - Go to MC, Enter the Blog ID then click “Open reportcard”.
  - In the “Account Rejection Tool” section.
  - Chose “Refund all” option then click “Reject Account”.
 
**b) Hardcode the account status**

- On the WCPay Server repo.
- Edit `server/wp-content/rest-api-plugins/endpoints/wcpay/class-accounts-controller.php` line `885`
- Change the status to: `'status' => 'rejected.something',`
- Note that this will reject all WCPay accounts. So make sure to discard the changes after you are done testing.

I'd suggest using option "b" if you don't want to go through the hassle of creating a new WCPay account every time you want to test a rejected account.

---

#### Disable WooPay for rejected accounts.

- In the WCPay Server’ admin area:
    - Reject your merchant account using the steps above.
- Go to the WC Pay site’ admin area
    - Go to “WCPay Dev“
    - Clear the "Account cache contents” to apply changes.
    - Go to > Payments. The account should be “Rejected”.
- Go to the Store
    - Add products.
    - Go to the Shortcode Cart page.
        - Check the Network tab in Chrome. It shouldn’t download any Woopay related file.
    - Go to the Cart Block page.
        - Check the Network tab in Chrome. It shouldn’t download any Woopay related file.
    - Go to Shortcode Checkout page.
        - Check the Network tab in Chrome. It shouldn’t download any Woopay related file.
    - Go to Checkout Block page.
        - Check the Network tab in Chrome. It shouldn’t download any Woopay related file.

#### Disable WooPay for suspended accounts.

- In the WCPay Server’ admin area:
    - Go to MC, Enter the Blog ID then click “Open reportcard”.
    - In the “Account Tool” section:
    - Select “Hard Block Account” then click “Apply Account Action”.
- Go to the WC Pay site’ admin area
    - Go to “WCPay Dev“
    - Clear the "Account cache contents” to apply changes.
    - Go to > Payments. The account should be “Under review”.
- Go to the Store
    - Add products.
    - Go to the Shortcode Cart page.
        - Check the Network tab in Chrome. It shouldn’t download any Woopay related file.
    - Go to the Cart Block page.
        - Check the Network tab in Chrome. It shouldn’t download any Woopay related file.
    - Go to Shortcode Checkout page.
        - Check the Network tab in Chrome. It shouldn’t download any Woopay related file.
    - Go to Checkout Block page.
        - Check the Network tab in Chrome. It shouldn’t download any Woopay related file.

#### Remove active WooPay webhooks for rejected accounts.

- Go to the WC Pay site’ admin area
    - Reject your merchant account using the steps above.
- Go to the WC Pay site’ admin area
    - Go to “WCPay Dev“
    - Clear the "Account cache contents” to apply changes.
    - Go to > Payments. The account should be “Rejected”.
    - Go to WooCommerce > Settings > Advanced > Webhooks
    - Verify there’s no “WCPay woopay order status sync” webhook.

#### Remove active WooPay webhooks for suspended accounts.

- Make sure to clear the WCPay account on the WC Pay Server using the Account Tool on the Report Card.
- Go to the WC Pay site’ admin area
    - Go to WooCommerce > Settings > Advanced > Webhooks
    - Verify there’s a “WCPay woopay order status sync” active webhook.
- In the WCPay Server’ admin area:
    - Go to MC, Enter the Blog ID then click “Open reportcard”.
    - In the “Account Tool” section:
    - Select “Hard Block Account” then click “Apply Account Action”.
- Go to the WC Pay site’ admin area
    - Go to “WCPay Dev“
    - Clear the "Account cache contents” to apply changes.
    - Go to > Payments. The account should be “Under review”.
    - Go to WooCommerce > Settings > Advanced > Webhooks
    - Verify there’s no “WCPay woopay order status sync” webhook.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

asumaran committed:
Merchant is not eligible if the account has been rejected

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8857` or branch name `as-disable-woopay-rejected-suspended-accounts` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=as-disable-woopay-rejected-suspended-accounts) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 21ba1276ee13dd160d24a2c7db5509fd9f3b200e
- Build time: 2024-05-30 22:53:07 UTC

_Note: the build is updated when a new commit is pushed to this PR._

asumaran committed:
Do not enable WooPay if the account is under review

asumaran committed:
Add changelog entry

asumaran committed:
Do not register WooPay hooks if the merchant is suspended or rejected.

Also removes active webhooks if the account is suspended or rejected

asumaran committed:
Check if the `platform_checkout_eligible` index exists before using it

asumaran committed:
Add tests to verify the `is_woopay_eligible` will return false if the
account is restriced

ricardo commented on includes/class-wc-payments-features.php:
```diff
@@ -210,7 +210,15 @@ public static function is_woopay_eligible() {
 		$account = WC_Payments::get_database_cache()->get( WCPay\Database_Cache::ACCOUNT_KEY, true );
-		return is_array( $account ) && ( $account['platform_checkout_eligible'] ?? false );
+
+		$is_account_rejected = WC_Payments::get_account_service()->is_account_rejected();
+
+		$is_account_under_review = WC_Payments::get_account_service()->is_account_under_review();
```
I created this issue only taking into account "rejected*" and "under_review" statuses, but it looks like there could be [more statuses](https://github.com/Automattic/woocommerce-payments-server/blob/9e3b5c29913af25da82e7efb7289e15d72a83ef8/server/wp-content/rest-api-plugins/endpoints/wcpay/stripe/class-stripe-account.php#L26-L31) and I'm not sure if/how these would be set in a real scenario.

@asumaran should we consider those other statuses as well? Perhaps untimately it would be easier to just check if the account is enabled/complete/not currently restricted?

asumaran replied:
> should we consider those other statuses as well? 

Good question. We'd need to confirm what statuses actually block the account. I'll double check tomorrow. In the meantime I'll ask internally about that.

asumaran replied:
I noticed the account status is set to `restricted` in two cases:

- When payments are paused on Stripe.
- When payouts are paused on Stripe.

When the merchant's account is `restricted` WooPayments is not disabled on the site neither the WooPayments account.

Does it make sense to turn off the merchant's WooPay eligibility if their Stripe account is restricted? cc @pierorocca @bborman22

asumaran replied:
The code on WooPayments is actually working as expected. I've removed most of the code I was going to introduce. 

The WooPay webhook is not registered if the account is suspended or rejected. Here in my PR I'm just removing the webhook in case the account is not eligible.

The question above still stands but it should be implemented in the [server](https://github.com/Automattic/woocommerce-payments-server/pull/5453) instead.

asumaran replied:
I've tested with a JN site + my sandbox and can confirm WooPay doesn't work if the merchant has payments disabled on Stripe. No iframe is triggered after entering a valid WooPay email. The account status is set to `restricted`.

The server will set the site as not eligible for WooPay. So, nothing to add there as it's already being handled correctly.

>Perhaps untimately it would be easier to just check if the account is enabled/complete/not currently restricted?

I think disabling WooPay for "rejected", "under review" and "restricted" accounts it's OK from WooPayments. If there's more status we need to handle it would be better to do it server side.

ricardo reviewed (approved):
Code LGTM and all tests have passed.

- ✅ Disable WooPay for rejected accounts.
- ✅ Disable WooPay for suspended accounts.
- ✅ Remove active WooPay webhooks for rejected accounts.
- ✅ Remove active WooPay webhooks for suspended accounts.

@asumaran Please check my comment below before merging.

asumaran committed:
Delay WooPay hooks registration

Fixes a fatal error when clearing the account data cache.

In this PR we call `WC_Payments::get_account_service()->is_account_rejected()` to know if the account is rejected. This method will get the data from Stripe and the `WC_Payments_Account->get_cached_account_data()` will be called. Then the action `woocommerce_payments_account_refreshed` will be fired.

The `Compatibility_Service` class is registering a hook for that action. https://github.com/Automattic/woocommerce-payments/blob/trunk/includes/class-compatibility-service.php#L41
Then `Compatibility_Service->get_compatibility_data()` is called. Which calls the `get_permalink()` function.

This function internally calls to `_get_page_link` which uses `$wp_rewrite` global but at the time it’s not defined resulting in a PHP fatal error. This commit fixes it.

Note that the request to the server is not done on every page refresh, only after clearing the account cache contents.

asumaran committed:
Remove unnecessary calls to `is_account_rejected` and `is_account_under_review`

We are already checking if the account is rejected or under review in the `WC_Payments_Features::is_woopay_eligible()` method.

asumaran commented on includes/class-wc-payments.php:
```diff
@@ -569,7 +569,9 @@ public static function init() {
 		wcpay_get_container()->get( \WCPay\Internal\Service\DuplicatePaymentPreventionService::class )->init_hooks();
 
-		self::maybe_register_woopay_hooks();
+		// Defer registering the WooPay hooks. Later on, $wp_rewrite is used and causes a fatal error on the WooPayment Dev Tools,
```
The fatal occurs every time the account cache is refreshed.

asumaran commented on includes/class-wc-payments.php:
```diff
@@ -569,7 +569,9 @@ public static function init() {
-		self::maybe_register_woopay_hooks();
+		// Defer registering the WooPay hooks. Later on, $wp_rewrite is used and causes a fatal error on the WooPayment Dev Tools,
+		// given that $wp_rewrite is defined right after the `plugins_loaded` action is fired. See #8857.
+		add_action( 'setup_theme', [ __CLASS__, 'maybe_register_woopay_hooks' ] );
```
It would be good to have more eyes on this. The reasoning is explained in the commit message https://github.com/Automattic/woocommerce-payments/pull/8857/commits/69fa9ef8a2b302287dc2c315a9ec8646fdf80f33 

I've tested the full WooPay flow locally and did not experience any errors with this change.

Pinging @waclawjacek since he introduce this line initially and @malithsen since worked on something [very similar](https://github.com/Automattic/woocommerce-payments/pull/8118).

waclawjacek replied:
Hi! :wave: Could you please provide some more details on how I could help here? I'm not sure what the code I introduced is, and I've been detached from this codebase for quite some time so some extra context would be much appreciated.

asumaran replied:
Hi @waclawjacek. I believe the original line was added [in this PR of yours](https://github.com/Automattic/woocommerce-payments/pull/3481/files#diff-57ee486c4a3ba4bae478eb31255c9a968b5b4eccd04b27f512e4003440d6f821R163). I just wanted to get more eyes on this from people familiar with it. The change looks a bit sensitive since I'm delaying some hooks registrations for the reasons mentioned in https://github.com/Automattic/woocommerce-payments/pull/8857/commits/4e1b201cd88b1445fa670698a522dd58e9dcb594. I've tested it with WooPay and double-checked that it's working. If you don't see anything to comment that's fine. I reached out to you as an extra precaution ;-)

waclawjacek replied:
I see! Thanks for the explanation and the ping - I appreciate the cautiousness.

I haven't tested the code or done a super deep dive into the WP codebase but the reasoning sounds good to me. It looks like [`setup_theme` is the next hook that fires in `wp-settings.php`](https://github.com/WordPress/wordpress-develop/blob/3a9ecd2cc84fe24e2db8b537a73314c5398b5312/src/wp-settings.php#L630) after `plugins_loaded` ([which is when `WC_Payments::init()` fires right now](https://github.com/Automattic/woocommerce-payments/blob/21ba1276ee13dd160d24a2c7db5509fd9f3b200e/woocommerce-payments.php#L175)), and it looks like a good candidate to hook into, despite the hook's name being somewhat irrelevant to the logic here.

malithsen replied:
Apologies for the delay in getting to this.

Change looks good to me. I tested WooPay locally with a WCPay account having `status=complete` and did not notice any regressions.

asumaran committed:
Make comment more accurate
